pkg/operation/botanist/migration.go

```diff
@@ -0,0 +1,168 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package botanist
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/gardener/gardener/pkg/api/extensions"
+	gardencorev1alpha1helper "github.com/gardener/gardener/pkg/apis/core/v1alpha1/helper"
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+	v1beta1constants "github.com/gardener/gardener/pkg/apis/core/v1beta1/constants"
+	extensionsv1alpha1 "github.com/gardener/gardener/pkg/apis/extensions/v1alpha1"
+	"github.com/gardener/gardener/pkg/client/kubernetes"
+	"github.com/gardener/gardener/pkg/operation/common"
+	"github.com/gardener/gardener/pkg/utils/flow"
+	kutil "github.com/gardener/gardener/pkg/utils/kubernetes"
+	"github.com/gardener/gardener/pkg/utils/retry"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	k8sretry "k8s.io/client-go/util/retry"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// AnnotateExtensionCRsForMigration annotates extension CRs with migrate operation annotation
+func (b *Botanist) AnnotateExtensionCRsForMigration(ctx context.Context) (err error) {
+	var fns []flow.TaskFn
+	fns, err = b.applyFuncToAllExtensionCRs(ctx, annotateObjectForMigrationFunc(ctx, b.K8sSeedClient.Client()))
+	if err != nil {
+		return err
+	}
+
+	return flow.Parallel(fns...)(ctx)
+}
+
+func annotateObjectForMigrationFunc(ctx context.Context, client client.Client) func(runtime.Object) error {
+	return func(obj runtime.Object) error {
+		return kutil.TryUpdate(ctx, k8sretry.DefaultBackoff, client, obj, func() error {
+			acc, err := extensions.Accessor(obj)
+			if err != nil {
+				return err
+			}
+
+			kutil.SetMetaDataAnnotation(acc, v1beta1constants.GardenerOperation, v1beta1constants.GardenerOperationMigrate)
+			return nil
+		})
+	}
+}
+
+// DeleteAllExtensionCRs deletes all extension CRs from the Shoot namespace
+func (b *Botanist) DeleteAllExtensionCRs(ctx context.Context) error {
+	fns, err := b.applyFuncToAllExtensionCRs(ctx, func(obj runtime.Object) error {
+		if err := common.ConfirmDeletion(ctx, b.K8sSeedClient.Client(), obj); err != nil {
+			return err
+		}
+
+		if err := client.IgnoreNotFound(b.K8sSeedClient.Client().Delete(ctx, obj, kubernetes.DefaultDeleteOptions...)); err != nil {
+			accObj, err := meta.Accessor(obj)
+			if err != nil {
+				return err
+			}
+			return fmt.Errorf("couldn't delete CR %s with name %s: %v", obj.GetObjectKind(), accObj.GetName(), err)
+		}
+		return nil
```

Comment:
Can you call https://github.com/gardener/gardener/blob/master/pkg/operation/common/extensions.go#L113 instead?